### ISSUE TITLE
AKS clusters should use availability zones for regions that support it

### ISSUE THREAD

**Issue by BernieWhite** [Azure/PSRule.Rules.Azure#880]
# Rule request

## Suggested rule change

AKS clusters support availability zones. Clusters deployed into regions that also support availability zones should have this configured.

## Applies to the following

The rule applies to the following:

- Resource type: **Microsoft.ContainerService/managedClusters**

**Comment by ArmaanMcleod** [comment]
@BernieWhite I'd be happy to help with this one as well 🙂.

Would this apply to pre-flight and in-flight data? Or just in-flight?

I also don't know of a reliable way to determine if a region supports Availability zones. I have used `Get-AzComputeResourceSku`  in the past to list VM skus with supported zones, which I guess could be used with AKS node sizes to determine if availability zones are supported for a region. 

```pwsh
> (Get-AzComputeResourceSku -Location 'australiaeast' | Where-Object { $_.Name -eq 'Standard_DS2_v2' -and $_.ResourceType -eq 'virtualMachines' }).LocationInfo
Location      Zones     ZoneDetails
--------      -----     -----------
australiaeast {1, 3, 2} {}
```

Perhaps there is a better API to check this?

Was also thinking we could just check if `availabilityZones` is `null` in the template, then notify users that availability zones should be used. If a region doesn't support availability zones, then the rule could be suppressed. Feel like this rule would get suppressed alot given regions like `australiasoutheast` do not support availability zones at all 😄 .

**Comment by BernieWhite** [comment]
@ArmaanMcleod This should be valid for pre-flight and in-flight data. My initial thought was that the supported AZ regions for AKS can be specified by configuration. Allowing this to be easily updated in future updates or by customers. Support for AZ doesn't change on a week-by-week basis.

The list [Limitations and region availability](https://docs.microsoft.com/en-us/azure/aks/availability-zones#limitations-and-region-availability) is fairly small.

In terms of dumping it, that's a good idea. The information should be expose via https://docs.microsoft.com/en-us/rest/api/resources/providers/list.

The list is based on AZ availability for `Microsoft.Compute/virtualMachineScaleSets` since using scale sets in cluster creation is a requirements.

**Comment by BernieWhite** [comment]
Additionally, in `data/` we include provider information, AZ information could be added here which is shipped in the module and can be looked up as the master list. Then provide a config value just for additional AZ locations instead of the master list.

**Comment by ArmaanMcleod** [comment]
Thanks @BernieWhite. I will add the AZ information for `Microsoft.Compute/virtualMachineScaleSets` to `providers.json`. 

Also would the configuration add AZ locations to the master list of completely override? I'd assume override would be simpler.

**Comment by BernieWhite** [comment]
> Thanks @BernieWhite Bernie White FTE. I will add the AZ information for `Microsoft.Compute/virtualMachineScaleSets` to `providers.json`.
> 
> Also would the configuration option add AZ locations to the master list or completely override? I'd assume override would be simpler.

I was thinking more add locations, however either way addresses the original intent of the rule.

The main call out that we need to test is that location names are normalized. Both these forms are valid:

- `eastus`
- `East US`

**Comment by ArmaanMcleod** [comment]
@BernieWhite Yeah I think add locations might be more convenient for user, I'll do that. 

For normalizing locations, I can probably use `GetNormalLocation` when doing the comparison.

**Comment by ArmaanMcleod** [comment]
@BernieWhite I'm wondering how this will work for standalone `Microsoft.ContainerService/managedClusters/agentPools` resource declared in the template:

```json
{
    "type": "Microsoft.ContainerService/managedClusters/agentPools",
    "apiVersion": "2019-10-01",
    "name": "[concat(parameters('clusterName3'), '/agentpool3')]",
     "dependsOn": [
         "[resourceId('Microsoft.ContainerService/managedClusters', parameters('clusterName3'))]"
     ],
     "properties": {
        "count": 3,
        "vmSize": "Standard_B2ms",
        "osDiskSizeGB": 100,
         "vnetSubnetID": "[concat(parameters('vnetId'), '/subnets/subnet-03')]",
         "maxPods": 50,
         "type": "VirtualMachineScaleSets",
         "orchestratorVersion": "1.20.5",
         "osType": "Linux",
         "enableAutoScaling": false,
         "availabilityZones": null
    }
}
```

Given this does not include `location`. Would the rule need to lookup the location from `name` or the `dependsOn` resource in this case?

**Comment by BernieWhite** [comment]
@ArmaanMcleod Good question.

The currently behaviour is: if the `managedClusters` and the `managedClusters/agentPools` are in the same template with a depends on then the expansion process will automatically nest the `managedClusters/agentPools` as a sub-resource for the `managedClusters` resource. You don't need to do anything for that.

So you could apply the location of the `managedClusters` in the `GetAgentPoolProfiles` helper function to any `managedClusters/agentPools`.

In edge case is if you deploy a `managedClusters/agentPools` in a separate template by itself, there is no effective way to address this so it should be excluded.
